### ISSUE TITLE
[Refactor] `SimulateCarry` to use `NamedTuple` for better readability

### ISSUE THREAD

**Issue by hanson-hschang** [hanson-hschang/Signal-System#66]
Currently, `SimulateCarry` is defined as a plain `tuple`. Passing and unpacking a plain tuple inside the `jax.lax.scan` body makes the code less readable and more fragile to structural changes. Having attribute access (e.g., `carry.time`, `carry.state`) would significantly improve code clarity.

https://github.com/hanson-hschang/Signal-System/blob/d89b9bfba89d605b3e92fe75c4910dcad58cb301/src/ss/system/_system.py#L105-L109

I propose redefining `SimulateCarry` as a `typing.NamedTuple`. JAX natively registers `NamedTuple` as a PyTree, meaning it works out-of-the-box with `lax.scan`, `jax.jit`, and auto-diff tools without any boilerplate.

```python
from typing import NamedTuple

class SimulateCarry(NamedTuple):
    time: float
    state: Shaped[Array, "batch_size ..."]
    controller_state: ControllerState | None

```

For the `body` inside the `simulate` loop, we can simply instantiate a new one at each step, which is clean and requires no special update methods:
```python
    @jax.jit
    def body(
        carry: SimulateCarry,
        random_key: PRNGKeyArray,
    ) -> tuple[SimulateCarry, SimulateStep]:
        # ... process logic ...
        next_carry = SimulateCarry(next_time, next_state, next_controller_state)
        return next_carry, (next_time, state, observation, control)
```

## Alternatives Considered & Comparison
Since JAX flattens PyTrees during compilation, all options below result in **identical XLA performance and identical gradient tracking** for differentiable filtering.

### 1. `equinox.Module`
Fits well if standardizing on Equinox, but requires importing `dataclasses.replace` for updates and makes `equinox` a strict dependency just for a data container.

```python
import equinox as eqx
from dataclasses import replace

class SimulateCarry(eqx.Module):
    time: float
    state: Array
    controller_state: ControllerState | None

# Update inside scan:
next_carry = replace(carry, time=next_time, state=next_state, controller_state=next_controller_state)

```

### 2. `flax.struct.dataclass`
Provides a very clean built-in `.replace()` method, but requires adding `flax` as a dependency, which is unnecessary overhead if the project otherwise uses Equinox or pure JAX.

```python
from flax import struct

@struct.dataclass
class SimulateCarry:
    time: float
    state: Array
    controller_state: ControllerState | None

# Update inside scan:
next_carry = carry.replace(time=next_time, state=next_state, controller_state=next_controller_state)

```

## Conclusion
Given that we update all three fields (`time`, `state`, `controller_state`) at every step of the simulation, the lightweight, native, and dependency-free nature of `NamedTuple` makes it the best fit.

**Comment by skim0119** [comment]
@hanson-hschang 
## Few minor thoughts:
- Using `flax` at this point doesn't make sense. We should stick with either `equinox` or `flax`, not both.
- Not a huge fan of `NamedTuple`. It feels like trying to write C/C++ style struct in python. I would prefer dataclass. In this case, `eqx.Module` does what `dataclass` should do.
- `typing.NamedTuple` is actually in `collections.NamedTuple`.

## Major thought:
I'm not too annoyed by having type-alias declaration instead of dataclass for `SimulateCarry`: at the end, it is just a short-name for `Carry` that is passed to the scan, and only for type checking. It is exactly doing the same thing as having layer of actual class definition. `Carry` type is only meant to be used during the `mypy` check, and excluded during the run. Using `dataclass` will have some overhead (I'm not sure how much `jit` will optimize, it might actually be the same).

> For the body inside the simulate loop, we can simply instantiate a new one at each step, which is clean and requires no special update methods:

I think it will just add a layer of instantiation and un-rolling around the block of code for scan, and code will still look a mess anyway. See:

```py
    @jax.jit
    def body(
        carry: SimulateCarry,
        random_key: PRNGKeyArray,
    ) -> tuple[SimulateCarry, SimulateStep]:
        time, state, controller_state = carry.time, carry.state, carry.controller_state
        # ... process logic ... (same for both)
        next_carry = SimulateCarry(next_time, next_state, next_controller_state)
        return next_carry, (next_time, state, observation, control)
```
vs
```py
    @jax.jit
    def body(
        carry: SimulateCarry,
        random_key: PRNGKeyArray,
    ) -> tuple[SimulateCarry, SimulateStep]:
        time, state, controller_state = carry
        # ... process logic ... (same for both)
        return (next_time, state, next_controller_state), \
               (next_time, state, observation, control)
```

For readability, I think it is about the same level, it is just matter of "do you read comments first" or "do you read code first":

```py
@struct.dataclass
class SimulateCarry:
    time: float
    state: Array
    controller_state: ControllerState | None
```
vs
```py
type SimulateCarry = tuple[ 
    float,  # time 
    Shaped[Array, "batch_size ..."],  # system state 
    ControllerState | None,  # controller state 
 ] 
```
I believe the type-alias is a bit more accurate in-terms of what this code is supposed to do.

## My verdict
If `carry` is simple enough, like just tuple of time (float) and few states, I think adding layer of class definition doesn't save much. If we start to pack more stuffs in `carry`, we can use `eqx.Module`.

**Comment by hanson-hschang** [comment]
@skim0119  
- I dont think the following is needed: `time, state, controller_state = carry.time, carry.state, carry.controller_state` inside the `body`. A direct call from `carry.xxx` for each place that is needed is enough, and it also provides a clear meaning that this is something carry through every loop.
- I think `time, state, controller_state = carry` this is prone to error if the order is changed or extra items are added.
- For readability and maintainability, I think a dataclass is better on both ends.
- In addition, Im thinking some future feature of saving data related to #65 :
```python
next_carry = SimulateCarry(next_time, next_state, next_controller_state)
next_carry.save_to(storage) # some method to save the status to the storage
```

and

```python
@dataclass
class SimulateCarry:
    time: float
    state: Array
    controller_state: ControllerState | None
    
    def save_to(self, storage: SequenceData) -> None:
        storage.save(time=self.time, state=self.state, controller_state=self.controller_state)
```

Also, I'm fine with both `eqx.Module` or `dataclass`.
